File: TemplateMethod/python/TemplateMethodPattern.py
class Template:

    def __init__(self):
        pass

    @staticmethod
    def wake_up():
        print("##### Wake Up by Alarm #####")

    def go2_work(self):
        pass

    def work_hard(self):
        pass

    def comeback2_home(self):
        pass

    @staticmethod
    def go2_sleep():
        print("***** Sweet Dream *****")

    def show_daily_life(self):
        self.wake_up()
        self.go2_work()
        self.work_hard()
        self.comeback2_home()
        self.go2_sleep()


class Developer(Template):

    def go2_work(self):
        print("@@ Please, no error today~~")

    def work_hard(self):
        print("@@@@@ What happened? It worked yesterday~~")

    def comeback2_home(self):
        print("@SuppressWarnings...")


class Student(Template):

    def go2_work(self):
        print("^^ Please, no exam today~~")

    def work_hard(self):
        print("^^^^^ Damn, I forgot again~~")

    def comeback2_home(self):
        print("^ Don't worry, I am still young...")


def run():
    # 어차피 동적 타입이라
    # 추상 템플릿 클래스 타입으로 사용되지 않으므로
    # 템플릿 메서드 패턴의 의미가 있나 모르겠..
    developer = Developer()
    developer.show_daily_life()

    student = Student()
    student.show_daily_life()


run()

File: TemplateMethod/ecma2015/TemplateMethodPattern.js
'use strict';
class Template {
    constructor() {}

    showDailyLife() {
        this.wakeUp();
        this.go2Work();
        this.workHard();
        this.comeback2Home();
        this.go2Sleep();
    }

    wakeUp() {
        console.log("##### Wake Up by Alarm #####");
    }

    go2Work() {
        throw Error("This method should be overridden");
    }

    workHard() {
        throw Error("This method should be overridden");
    }

    comeback2Home() {
        throw Error("This method should be overridden");
    }

    go2Sleep() {
        console.log("***** Sweet Dream *****");
    }
}

class Developer extends Template {
    constructor() {
        super();
    }

    go2Work() {
        console.log("@@ Please, no error today~~");
    }

    workHard() {
        console.log("@@@@@ What happened? It worked yesterday~~");
    }

    comeback2Home() {
        console.log("@SuppressWarnings...");
    }
}

class Student extends Template {
    constructor() {
        super();
    }

    go2Work() {
        console.log("^^ Please, no exam today~~");
    }

    workHard() {
        console.log("^^^^^ Damn, I forgot again~~");
    }

    comeback2Home() {
        console.log("^ Don't worry, I am still young...");
    }
}

(() => {
    "use strict";

    console.log("======= Template for Developer =======");
    let developer = new Developer();
    developer.showDailyLife();

    console.log('\n');

    console.log("======= Template for Student =======");
    let student = new Student();
    student.showDailyLife();

})();